tajo-core/src/main/java/org/apache/tajo/engine/function/hiveudf/HiveFunctionLoader.java

```diff
@@ -0,0 +1,161 @@
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.ql.exec.Description;
+import org.apache.hadoop.hive.ql.exec.UDF;
+import org.apache.hadoop.hive.ql.udf.UDFType;
+import org.apache.hadoop.io.Writable;
+import org.apache.tajo.catalog.FunctionDesc;
+import org.apache.tajo.catalog.FunctionDescBuilder;
+import org.apache.tajo.catalog.proto.CatalogProtos;
+import org.apache.tajo.common.TajoDataTypes;
+import org.apache.tajo.conf.TajoConf;
+import org.apache.tajo.exception.TajoInternalError;
+import org.apache.tajo.function.UDFInvocationDesc;
+import org.apache.tajo.util.WritableTypeConverter;
+import org.reflections.Reflections;
+import org.reflections.util.ConfigurationBuilder;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.*;
+
+public class HiveFunctionLoader {
+  public static Optional<List<FunctionDesc>> loadHiveUDFs(TajoConf conf) {
+    ArrayList<FunctionDesc> funcList = new ArrayList<>();
+    String udfdir = conf.getVar(TajoConf.ConfVars.HIVE_UDF_DIR);
+
+    try {
+      Path udfPath = new Path(udfdir);
+      FileSystem fs = udfPath.getFileSystem(conf);
+
+      if (!fs.isDirectory(udfPath)) {
+        return Optional.empty();
+      }
+
+      // loop each jar file
+      for (FileStatus fstatus : fs.listStatus(udfPath, (Path path) -> path.getName().endsWith(".jar"))) {
+
+        URL[] urls = new URL[]{new URL("jar:" + fstatus.getPath().toUri().toURL() + "!/")};
+
+        // extract and register UDF's decendants (legacy Hive UDF form)
+        Set<Class<? extends UDF>> udfClasses = getSubclassesFromJarEntry(urls, UDF.class);
+        if (udfClasses != null) {
+          buildFunctionsFromUDF(udfClasses, funcList, "jar:"+urls[0].getPath());
+        }
+      }
+    } catch (IOException e) {
+      throw new TajoInternalError(e);
+    }
+
+    return Optional.of(funcList);
+  }
+
+  private static <T> Set<Class<? extends T>> getSubclassesFromJarEntry(URL[] urls, Class<T> targetCls) {
+    Reflections refl = new Reflections(new ConfigurationBuilder().
+        setUrls(urls).
+        addClassLoader(new URLClassLoader(urls)));
+
+    return refl.getSubTypesOf(targetCls);
+  }
+
+  static void buildFunctionsFromUDF(Set<Class<? extends UDF>> classes, List<FunctionDesc> list, String jarurl) {
+    for (Class<? extends UDF> clazz: classes) {
+      String [] names;
+      String value = null, extended = null;
+
+      Description desc = clazz.getAnnotation(Description.class);
+
+      // Check @Description annotation (if exists)
+      if (desc != null) {
+        names = desc.name().split(",");
+        for (int i=0; i<names.length; i++) {
+          names[i] = names[i].trim();
+        }
+
+        value = desc.value();
+        extended = desc.extended();
+      }
+      else {
+        names = new String [] {clazz.getName().replace('.','_')};
```

Comment:
This part is from Hive code and I guess it is to avoid name conflict.
Keeping it is not bad in my opinion because it can force to use the annotation.